torch/csrc/generic/StorageMethods.cpp

```diff
@@ -102,38 +102,49 @@ static PyObject * THPStorage_(fromBuffer)(PyObject *_unused, PyObject *args, PyO
   PyObject *obj = nullptr;
   const char* byte_order_str = nullptr;
   Py_ssize_t count = -1, offset = 0;
+  PyObject* dtype_obj = nullptr;
+  c10::ScalarType scalar_type = at::kByte;
   Py_buffer buffer = {};
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays,clang-diagnostic-writable-strings)
-  static char *kwlist[] = {"buffer", "byte_order", "count", "offset", nullptr};
+  static char *kwlist[] = {"buffer", "byte_order", "count", "offset", "dtype", nullptr};
   // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
   const char* argtypes;
-#if defined(TH_REAL_IS_BYTE) || defined(TH_REAL_IS_CHAR)
-  argtypes = "O|snn";
-#else
-  argtypes = "Os|nn";
-#endif
+  argtypes = "O|snnO";
 
   if (!PyArg_ParseTupleAndKeywords(args, keywds, argtypes, kwlist,
-        &obj, &byte_order_str, &count, &offset)) {
+        &obj, &byte_order_str, &count, &offset, &dtype_obj)) {
     return nullptr;
   }
 
-#if !(defined(TH_REAL_IS_BYTE) || defined(TH_REAL_IS_CHAR))
+  if (dtype_obj != nullptr) {
```

Comment:
Wondering if it's safer to require `dtype_obj` to be specified explicitly, so we don't miss use sites.

Reply:
Yeah good idea, I'll do that

Reply:
It's worth noting though, I've disallowed the `dtype` arg for `TypedStorage.from_buffer`, because it will just call `ByteStorage.from_buffer` with it's own dtype. Maybe it's odd to require `dtype` for `ByteStorage.from_buffer`, but throw an error if it's given for `TypedStorage.from_buffer`? I'm not sure

Reply:
I guess this suggests that it was a mistake to special case ByteStorage and we should have just made it a TypedStorage too. Dunno how hard this is to fix though.

Reply:
This function no longer depends on TH_REAL_IS_BYTE etc that are defined by THGenerate*Types.h, and therefore no longer belongs in generic (files in generic are compiled multiple times for the different types). The only instantiation that you need is THPByteStorage_fromBuffer, so you should move out this function to non-generic Storage.cpp. 
Same for THPStorage_(nbytes), right? It's only ever called as THPByteStorage_nbytes, and does not require generic specification of compile-time `scalar_t`, `TH_REAL_IS_BYTE` etc,

Reply:
> I guess this suggests that it was a mistake to special case ByteStorage and we should have just made it a TypedStorage too. Dunno how hard this is to fix though.

You might be right. I'm not entirely sure how long that would take either

Reply:
At least for this one, I think it's worth making an attempt right now, since it will avoid an inconsistency between untyped bytestorage and the rest of the typed storages.

Reply:
Yeah I agree. I'll start on it now

Reply:
Ok, so it turns out that this isn't very hard. In my local checkout, I've changed the special `ByteStorage` to `UntypedStorage` in all the appropriate places, and added a new `ByteStorage` based upon `TypedStorage`, just like all the other `<type>Storage`s. It works fine, though there are a couple small kinks to work out (like places in C++ land where I special cased the byte dtype). Probably should add a few tests for `UntypedStorage` as well.

I just had some questions on naming and API compatibility:

Is `UntypedStorage` alright, or is there a better name? I like `UntypedStorage` because it clearly differentiates itself from `TypedStorage` and all the `<type>Storage`s.

Should we allow `UntypedStorage` to be provided as an argument to things like `torch.tensor(storage)`, or should we disable those for now to be safe?

(One thing that might be worth mentioning is that `UntypedStorage` needs to be available as either `torch.UntypedStorage` for CPU or `torch.cuda.UntypedStorage` for CUDA, until we can cut out the `torch.cuda` storages in the future)

Reply:
@kurtamohler Apologies for the comment from a random onlooker, but I've been following this PR with interest as I have been waiting for quite some time for the ability to "reinterpret_cast" tensors (a la https://github.com/pytorch/pytorch/issues/29013). My particular use case is writing bindings for C++ video decoders.

> Should we allow UntypedStorage to be provided as an argument to things like torch.tensor(storage), or should we disable those for now to be safe?

If this question boils down to whether things like https://github.com/pytorch/pytorch/pull/59671#issuecomment-857923652 are possible, please do allow this. I would absolutely love the ability to create untyped storage, populate that storage in C++-land (by passing the pointer), and then interpret it as any type of tensor afterwards in Python-land.

Reply:
Thanks for the input @anibali! I've added `UntypedStorage`, and they can be provided as an argument to tensor methods